### PR TITLE
 Initial version of beatless

### PR DESCRIPTION
**Pull request by ph**
## This PR provide the following

1. Plugin infra for developing providers and functions
2. A local stdin provider only used for testing, I will remove it in the
final version.
3. AWS provider and function types for:
  - Cloudwatch logs
  - SQS
  - Kinesis
  - Api web gateway proxy
4. License checker
5. Packaging of artifact
6. Runners
7. CLI infrastructure
8. CLI to push a cloudwatch logs function.
9. CLI to delete any function
10. Processors support.
11. Types to validate value from the users and the lambda function.

## What it doesn't provides

- ECS and full event extraction. (NOT for v1)
- Specifying the AWS credentials in the configuration
- CLI for SQS, Kinesis, API
- Robust CLI interaction with the API, rollback on failure / versioning.
- Removal of not supported outputs
- Removal of seccomp check
- Integration tests
- Updated build task to produce containing the user executable beat and
the linux beats.
- Concurrency / memory settings

## Vendored

- aws/aws-lambda-go
- aws/aws-sdk-go-v2

---

**Note:** The PR look much bigger because of the vendored files.

**Notes 2:** 

Make sure you add this to your config YAMl, I wan't to make the following the default options but it would have been really messy to do it in this PR.

```
# Disable anything that could be send to disk
path.data: /tmp
path.logs: /tmp/logs
logging.to_stderr: true
logging.to_files: false
#logging.level: debug
setup.template.enabled: true
queue.mem:
  events: 50
  flush.min_events: 1
  flush.timeout: 0.5s
```

## related work
- https://github.com/elastic/beats/pull/8449
- https://github.com/elastic/beats/pull/8443

### CODE REVIEW

**Comment by ph** [comment]
Forgot about the license check, working on fixing that.

File: x-pack/beatless/beater/beatless.go [hunk 1]
```diff
@@ -5,68 +5,156 @@
...
-	done   chan struct{}
-	config config.Config
-	log    *logp.Logger
+var (
+	graceDelay   = 45 * time.Minute
+	refreshDelay = 15 * time.Minute
+)
 
-	// TODO: Add registry reference here.
+// Beatless is a beat designed to un under a serverless environment and listen to external triggers,
```
Comment (kvch) [inline x-pack/beatless/beater/beatless.go]: tyop: "to run"

File: x-pack/beatless/beater/beatless.go [hunk 2]
```diff
@@ -5,68 +5,156 @@
...
+// each invocation will generate one or more events to Elasticsearch.
+//
+// Each serverless implementation is different but beatless follows a few executions rules.
```
Comment (kvch) [inline x-pack/beatless/beater/beatless.go]: tyop: execution

File: x-pack/beatless/beater/beatless.go [hunk 3]
```diff
@@ -5,68 +5,156 @@
...
 }
 
 // New creates an instance of beatless.
 func New(b *beat.Beat, cfg *common.Config) (beat.Beater, error) {
-	c := config.DefaultConfig
-	if err := cfg.Unpack(&c); err != nil {
-		return nil, fmt.Errorf("error reading config file: %v", err)
+	c := &config.DefaultConfig
+	if err := cfg.Unpack(c); err != nil {
+		return nil, fmt.Errorf("xerror reading config file: %v", err)
```
Comment (kvch) [inline x-pack/beatless/beater/beatless.go]: xerror? is this a typo?

File: x-pack/beatless/core/coordinator.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+func NewCoordinator(log *logp.Logger,
+	runners ...Runner,
+) *Coordinator {
+	if log == nil {
+		log = logp.NewLogger("coordinator")
+	}
+	return &Coordinator{log: log, runners: runners}
+}
+
+// Start starts each functions into an independant goroutine and wait until all the goroutine are
```
Comment (kvch) [inline x-pack/beatless/core/coordinator.go]: tyop: independent

File: x-pack/beatless/core/coordinator.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+// stopped to exit.
+func (r *Coordinator) Start(ctx context.Context) error {
```
Comment (kvch) [inline x-pack/beatless/core/coordinator.go]: I would rather name this function `Run`. `Start` implies there is a `Stop` function to call.

**Comment by ruflin** [comment]
@ph One idea to make this more reviewable and not have to tell Github which files it should expand and which ones not: Could you open a PR with just the dependency against the feature branch and then rebase this one on top of it when it's merged?

File: x-pack/beatless/provider/registry.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+// Errors generated by the registry when we are retrieving providers or functions from the main registry.
+var (
+	errInvalidProvider     = errors.New("invalid provider name")
+	errInvalidFunctionName = errors.New("invalid function name")
+	errInvalidType         = errors.New("incomptible type received for the feature")
+)
+
+// namespace is the namespace were providers will be registered in the global registry.
+var namespace = "beatless.provider"
```
Comment (kvch) [inline x-pack/beatless/provider/registry.go]: This could be `const`.

File: x-pack/beatless/provider/feature.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+// you may not use this file except in compliance with the Elastic License.
+
+package provider
+
+import "github.com/elastic/beats/libbeat/feature"
+
+// ns return the namespace for functions of a specific provider. The registry have a flat view
+// representation of the plugin world this mean we don't really have a tree, instead what we do is
+// to create a unique keys per providers that will only keep the functions of the provider.
+func ns(provider string) string {
```
Comment (kvch) [inline x-pack/beatless/provider/feature.go]: Please rename this function to `namespace` or something more descriptive.

File: x-pack/beatless/provider/aws/cloudwatch_logs.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+func (c *CloudwatchLogs) Run(_ context.Context, client core.Client) error {
+	lambda.Start(func(request events.CloudwatchLogsEvent) error {
+		parsedEvent, err := request.AWSLogs.Parse()
+		if err != nil {
+			c.log.Errorf("could not parse events from cloudwatch logs, error: %s", err)
+			return err
+		}
+
+		// defensive checks
+		if len(parsedEvent.LogEvents) == 0 {
```
Comment (kvch) [inline x-pack/beatless/provider/aws/cloudwatch_logs.go]: Why is it invalid to get no events?

**Comment by ph** [comment]
@ruflin @kvch  I've created https://github.com/elastic/beats/pull/8483 with only the vendor changes lets get that merged and I will rebase this one.

File: x-pack/beatless/provider/aws/cloudwatch_logs.go (hunk 8 above)
Comment (ph) [reply x-pack/beatless/provider/aws/cloudwatch_logs.go]: I've looked at the SDK; this cannot happen the function should not be invoked in that case. I've removed the defensive check for all the function types.

**Comment by ph** [comment]
Rebased + first round of comments fixed, I am waiting on merging #8483

**Pull request closed without merging**